src/lib/query/pci.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use crate::{Iface, NisporError};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(into = "String")]
+#[serde(try_from = "String")]
+pub struct PciAddress {
+    pub domain: u32,
+    pub bus: u8,
+    pub slot: u8,
+    pub function: u8,
+}
+
+impl std::fmt::Display for PciAddress {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}:{:02x}:{:02x}.{}",
+            if self.domain > u16::MAX.into() {
+                format!("{:08x}", self.domain)
+            } else {
+                format!("{:04x}", self.domain)
+            },
+            self.bus,
+            self.slot,
+            self.function
+        )
+    }
+}
+
+impl TryFrom<String> for PciAddress {
+    type Error = NisporError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::from_str(value.as_str())
+    }
+}
+
+impl From<PciAddress> for String {
+    fn from(v: PciAddress) -> String {
+        v.to_string()
+    }
+}
+
+impl FromStr for PciAddress {
+    type Err = NisporError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut items: Vec<&str> = s.split(":").collect();
+        let err = NisporError::invalid_argument(format!(
+            "Invalid PCI address, expecting string \
+                like 00:1b.0 or 0000:07:00.1, but got {s}"
+        ));
+        match items.len() {
+            3 | 2 => {
+                // It is safe to unwrap as we already checked the length.
+                let slot_function_str: Vec<&str> =
+                    items.pop().unwrap().split(".").collect();
+                if slot_function_str.len() != 2 {
+                    return Err(err.clone());
+                }
+
+                let slot = u8::from_str_radix(slot_function_str[0], 16)
+                    .map_err(|_| err.clone())?;
```

Comment:
You can avoid creating the error beforehand and doing err.clone() all the time. Make the `match` to return an `Option<PciAddress>`. Then, if you got `None`, create the error and return it.
Now it always create/allocate the error once. If there is a real error it does a clone.
In the other way it doesn't do anything if there is no error, and it creates the error once if there is an error.

Reply:
Created a function `fn parse_pci_addr(pci_addr_str: &str) -> Option<PciAddress>` allowing ? to work there.